handlers/secrets_test.go

```diff
@@ -0,0 +1,236 @@
+package handlers
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/swarm"
+	"github.com/openfaas/faas/gateway/requests"
+)
+
+func genFakeSecret(name string, data string, includeOwnerLabel bool) swarm.Secret {
+	secret := swarm.Secret{
+		ID: name,
+		Spec: swarm.SecretSpec{
+			Data: []byte(data),
+			Annotations: swarm.Annotations{
+				Name: name,
+			},
+		},
+	}
+
+	if includeOwnerLabel {
+		secret.Spec.Annotations.Labels = map[string]string{
+			ownerLabel: ownerLabelValue,
+		}
+	}
+
+	return secret
+}
+
+func getInitialSecrets(includeNonOwnedSecrets bool) map[string]swarm.Secret {
+	initialSecrets := map[string]swarm.Secret{
+		"foo":        genFakeSecret("foo", "baz", true),
+		"foobar":     genFakeSecret("foobar", "bar", true),
+		"baz_fuz":    genFakeSecret("baz_fuz", "foo", true),
+		"barfoo_baz": genFakeSecret("barfoo_baz", "foobar", true),
+	}
+
+	if includeNonOwnedSecrets {
+		initialSecrets["without_label"] = genFakeSecret("without_label", "test", false)
+	}
+
+	return initialSecrets
+}
+
+type fakeDockerSecretAPIClient struct {
+	secrets map[string]swarm.Secret
+}
+
+func newFakeDockerSecretAPIClient() fakeDockerSecretAPIClient {
+	return fakeDockerSecretAPIClient{
+		secrets: getInitialSecrets(true),
+	}
+}
+
+func (c *fakeDockerSecretAPIClient) Reset() {
+	c.secrets = getInitialSecrets(true)
+}
+
+func (c *fakeDockerSecretAPIClient) SecretList(
+	_ context.Context,
+	options types.SecretListOptions,
+) ([]swarm.Secret, error) {
+	secrets := []swarm.Secret{}
+
+	for _, value := range c.secrets {
+		secrets = append(secrets, value)
+	}
+
+	return secrets, nil
+}
+
+func (c *fakeDockerSecretAPIClient) SecretCreate(
+	_ context.Context,
+	secretDesc swarm.SecretSpec,
+) (types.SecretCreateResponse, error) {
+	id := secretDesc.Name
+
+	newSecret := swarm.Secret{
+		ID: id,
+		Spec: swarm.SecretSpec{
+			Annotations: swarm.Annotations{
+				Name: secretDesc.Name,
+				Labels: map[string]string{
+					ownerLabel: ownerLabelValue,
+				},
+			},
+			Data: secretDesc.Data,
+		},
+	}
+
+	c.secrets[secretDesc.Name] = newSecret
+
+	return types.SecretCreateResponse{ID: id}, nil
+}
+
+func (c *fakeDockerSecretAPIClient) SecretRemove(_ context.Context, id string) error {
+	if _, ok := c.secrets[id]; !ok {
+		return fmt.Errorf("secret with id: %s not found", id)
+	}
+
+	delete(c.secrets, id)
+
+	return nil
+}
+
+func (c *fakeDockerSecretAPIClient) SecretInspectWithRaw(
+	_ context.Context,
+	name string,
+) (swarm.Secret, []byte, error) {
+	return swarm.Secret{}, nil, nil
```

Comment:
This should return the secret specified by `name` _or_ it should return a `SecretInspectWithRaw Not Implemented` error, instead of an empty secret struct

Reply:
Done